<a href="https://colab.research.google.com/github/abdel2ty/IntenseAI_Notebooks_v1/blob/main/notebook3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<div style="background-color:#28a745; padding:15px; border-radius:10px; text-align:center; display:flex; justify-content:center; align-items:center; gap:12px;">
    <img src="attachment:fb8a8d7c-bce6-474e-a5fe-819eac5d187f.jpg" alt="Evergreen Logo" style="width:70px; height:70px; border-radius:50%;">
    <h1 style="color:white; font-size:40px; font-weight:bold; margin:0;">
        Evergreen.ai
    </h1>

<div style="margin-left:auto; padding-right:10px;">
        <h2 style="color:white; font-size:40px; font-weight:600; margin:0;">
            Pandas
        </h2>
</div>
</div>


# 🐼 **Pandas Interactive Tutorial**

## **1️⃣ Importing Pandas**

In [ ]:
import pandas as pd

# Load the dataset
#df = pd.read_csv("Covid19-counties-2023.csv")
df = pd.read_csv("arab_covid_dataset.csv")
df

date      county    state     fips  cases  deaths
0        1/1/2023       Cairo    Egypt   1001.0  18961   230.0
1        1/1/2023        Giza    Egypt   1003.0  67496   719.0
2        1/1/2023  Alexandria    Egypt   1005.0   7027   111.0
3        1/1/2023       Aswan    Egypt   1007.0   7692   108.0
4        1/1/2023       Asyut    Egypt   1009.0  17731   260.0
...           ...         ...      ...      ...    ...     ...
267004  3/23/2023  Sweetwater  Wyoming  56037.0  12519   139.0
267005  3/23/2023       Teton  Wyoming  56039.0  12150    16.0
267006  3/23/2023       Uinta  Wyoming  56041.0   6416    43.0
267007  3/23/2023    Washakie  Wyoming  56043.0   2700    51.0
267008  3/23/2023      Weston  Wyoming  56045.0   1906    23.0

[267009 rows x 6 columns]

## **2️⃣ Quick Numeric Insights**
Use **`.mean()`**, **`.median()`**, and **`.std()`** to get insights into numerical columns.

In [ ]:
df['cases'].mean()   # Average cases

np.float64(31464.733143826612)

In [ ]:
df['cases'].median() # Median cases 1 2 3 4  -- 1 2 3 4 5

7955.0

In [ ]:
df['cases'].std()    # Standard deviation

119311.92395520568

## **3️⃣ Exploring Categorical Columns**
- Use `.unique()` to list unique values
- Use `.value_counts()` to count occurrences

In [ ]:
df

date      county    state     fips  cases  deaths
0        1/1/2023       Cairo    Egypt   1001.0  18961   230.0
1        1/1/2023        Giza    Egypt   1003.0  67496   719.0
2        1/1/2023  Alexandria    Egypt   1005.0   7027   111.0
3        1/1/2023       Aswan    Egypt   1007.0   7692   108.0
4        1/1/2023       Asyut    Egypt   1009.0  17731   260.0
...           ...         ...      ...      ...    ...     ...
267004  3/23/2023  Sweetwater  Wyoming  56037.0  12519   139.0
267005  3/23/2023       Teton  Wyoming  56039.0  12150    16.0
267006  3/23/2023       Uinta  Wyoming  56041.0   6416    43.0
267007  3/23/2023    Washakie  Wyoming  56043.0   2700    51.0
267008  3/23/2023      Weston  Wyoming  56045.0   1906    23.0

[267009 rows x 6 columns]

In [ ]:
df['state'].unique()     # List of unique states

array(['Egypt', 'Saudi Arabia', 'United Arab Emirates', 'Morocco',
       'Algeria', 'Tunisia', 'Libya', 'Sudan', 'Iraq', 'Syria', 'Jordan',
       'Palestine', 'Lebanon', 'Qatar', 'Bahrain', 'Kuwait', 'Oman',
       'Mauritania', 'Somalia', 'Comoros', 'Djibouti', 'Yemen',
       'California', 'Colorado', 'Connecticut', 'Delaware',
       'District of Columbia', 'Florida', 'Georgia', 'Guam', 'Hawaii',
       'Idaho', 'Illinois', 'Indiana', 'Iowa', 'Kansas', 'Kentucky',
       'Louisiana', 'Maine', 'Maryland', 'Massachusetts', 'Michigan',
       'Minnesota', 'Mississippi', 'Missouri', 'Montana', 'Nebraska',
       'Nevada', 'New Hampshire', 'New Jersey', 'New Mexico', 'New York',
       'North Carolina', 'North Dakota', 'Northern Mariana Islands',
       'Ohio', 'Oklahoma', 'Oregon', 'Pennsylvania', 'Puerto Rico',
       'Rhode Island', 'South Carolina', 'South Dakota', 'Tennessee',
       'Texas', 'Utah', 'Vermont', 'Virgin Islands', 'Virginia',
       'Washington', 'West Virginia', 'W

In [ ]:

print("✅ Default value counts:")
print(df['state'].value_counts())

print("\n✅ normalize=True (proportions instead of counts):")
print(df['state'].value_counts(normalize=True))

print("\n✅ sort=False (keep original order of appearance):")
print(df['state'].value_counts(sort=False))

print("\n✅ ascending=True (sort counts in ascending order):")
print(df['state'].value_counts(ascending=True))

print("\n✅ dropna=False (include NaN in the counts):")
print(df['state'].value_counts(dropna=False))


✅ Default value counts:
state
Texas         20975
Georgia       13120
Virginia      10973
Kentucky       9912
Missouri       9594
              ...  
Qatar             5
Mauritania        5
Bahrain           5
Djibouti          5
Comoros           4
Name: count, Length: 78, dtype: int64

✅ normalize=True (proportions instead of counts):
state
Texas         0.078555
Georgia       0.049137
Virginia      0.041096
Kentucky      0.037122
Missouri      0.035931
                ...   
Qatar         0.000019
Mauritania    0.000019
Bahrain       0.000019
Djibouti      0.000019
Comoros       0.000015
Name: proportion, Length: 78, dtype: float64

✅ sort=False (keep original order of appearance):
state
Egypt                     27
Saudi Arabia              13
United Arab Emirates       7
Morocco                   12
Algeria                   12
                        ... 
Alabama                 5429
Alaska                  2268
American Samoa            81
Arizona                 1215
Arkansas  

In [ ]:
df['county'].unique().tolist()     # List of unique county

['Cairo',
 'Giza',
 'Alexandria',
 'Aswan',
 'Asyut',
 'Beheira',
 'Beni Suef',
 'Dakahlia',
 'Damietta',
 'Fayoum',
 'Gharbia',
 'Ismailia',
 'Kafr El Sheikh',
 'Luxor',
 'Matrouh',
 'Minya',
 'Monufia',
 'New Valley',
 'North Sinai',
 'Port Said',
 'Qalyubia',
 'Qena',
 'Red Sea',
 'Sharqia',
 'Sohag',
 'South Sinai',
 'Suez',
 'Riyadh',
 'Makkah',
 'Madinah',
 'Eastern Province',
 'Qassim',
 'Hail',
 'Tabuk',
 'Al Bahah',
 'Asir',
 'Najran',
 'Jazan',
 'Northern Borders',
 'Al Jouf',
 'Abu Dhabi',
 'Dubai',
 'Sharjah',
 'Ajman',
 'Fujairah',
 'Ras Al Khaimah',
 'Umm Al Quwain',
 'Casablanca',
 'Rabat',
 'Fes',
 'Marrakech',
 'Tangier',
 'Agadir',
 'Tetouan',
 'Meknes',
 'Oujda',
 'Kenitra',
 'Nador',
 'Safi',
 'Algiers',
 'Oran',
 'Constantine',
 'Annaba',
 'Blida',
 'Batna',
 'Sétif',
 'Béjaïa',
 'Tlemcen',
 'Mostaganem',
 'Ghardaïa',
 'Skikda',
 'Tunis',
 'Sfax',
 'Sousse',
 'Gabes',
 'Bizerte',
 'Gafsa',
 'Kairouan',
 'Mahdia',
 'Kasserine',
 'Kebili',
 'Tozeur',
 'Medenine',
 'T

In [ ]:
for i in df['county'].unique():
    print(i)

Cairo
Giza
Alexandria
Aswan
Asyut
Beheira
Beni Suef
Dakahlia
Damietta
Fayoum
Gharbia
Ismailia
Kafr El Sheikh
Luxor
Matrouh
Minya
Monufia
New Valley
North Sinai
Port Said
Qalyubia
Qena
Red Sea
Sharqia
Sohag
South Sinai
Suez
Riyadh
Makkah
Madinah
Eastern Province
Qassim
Hail
Tabuk
Al Bahah
Asir
Najran
Jazan
Northern Borders
Al Jouf
Abu Dhabi
Dubai
Sharjah
Ajman
Fujairah
Ras Al Khaimah
Umm Al Quwain
Casablanca
Rabat
Fes
Marrakech
Tangier
Agadir
Tetouan
Meknes
Oujda
Kenitra
Nador
Safi
Algiers
Oran
Constantine
Annaba
Blida
Batna
Sétif
Béjaïa
Tlemcen
Mostaganem
Ghardaïa
Skikda
Tunis
Sfax
Sousse
Gabes
Bizerte
Gafsa
Kairouan
Mahdia
Kasserine
Kebili
Tozeur
Medenine
Tataouine
Tripoli
Benghazi
Misrata
Sabha
Zliten
Derna
Sirte
Zawiya
Al Bayda
Khartoum
North Darfur
South Darfur
West Darfur
Gezira
Kassala
Blue Nile
Sennar
Baghdad
Basra
Nineveh
Anbar
Dhi Qar
Maysan
Wasit
Babil
Karbala
Najaf
Diyala
Kirkuk
Sulaymaniyah
Erbil
Dohuk
Salah al-Din
Al-Qadisiyyah
Muthanna
Damascus
Aleppo
Homs
Hama
Latakia
Ta

In [ ]:
df['state'].value_counts()

state
Texas         20975
Georgia       13120
Virginia      10973
Kentucky       9912
Missouri       9594
              ...  
Qatar             5
Mauritania        5
Bahrain           5
Djibouti          5
Comoros           4
Name: count, Length: 78, dtype: int64

## **4️⃣ Transforming Data Using `.map()` and `.apply()`**
- `.map()` → Apply a function to **Series**
- `.apply()` → Apply a function to **DataFrame or Series**

In [ ]:
df.head(2)

date county  state    fips  cases  deaths
0  1/1/2023  Cairo  Egypt  1001.0  18961   230.0
1  1/1/2023   Giza  Egypt  1003.0  67496   719.0

In [ ]:
# Example: Convert cases to thousands
df['cases_in_k'] = df['cases'] / 1000

In [ ]:
df

date      county    state     fips  cases  deaths  cases_in_k
0        1/1/2023       Cairo    Egypt   1001.0  18961   230.0      18.961
1        1/1/2023        Giza    Egypt   1003.0  67496   719.0      67.496
2        1/1/2023  Alexandria    Egypt   1005.0   7027   111.0       7.027
3        1/1/2023       Aswan    Egypt   1007.0   7692   108.0       7.692
4        1/1/2023       Asyut    Egypt   1009.0  17731   260.0      17.731
...           ...         ...      ...      ...    ...     ...         ...
267004  3/23/2023  Sweetwater  Wyoming  56037.0  12519   139.0      12.519
267005  3/23/2023       Teton  Wyoming  56039.0  12150    16.0      12.150
267006  3/23/2023       Uinta  Wyoming  56041.0   6416    43.0       6.416
267007  3/23/2023    Washakie  Wyoming  56043.0   2700    51.0       2.700
267008  3/23/2023      Weston  Wyoming  56045.0   1906    23.0       1.906

[267009 rows x 7 columns]

In [ ]:
df[['state', 'cases', 'cases_in_k']].head()

state  cases  cases_in_k
0  Egypt  18961      18.961
1  Egypt  67496      67.496
2  Egypt   7027       7.027
3  Egypt   7692       7.692
4  Egypt  17731      17.731

In [ ]:
df

date      county    state     fips  cases  deaths  cases_in_k
0        1/1/2023       Cairo    Egypt   1001.0  18961   230.0      18.961
1        1/1/2023        Giza    Egypt   1003.0  67496   719.0      67.496
2        1/1/2023  Alexandria    Egypt   1005.0   7027   111.0       7.027
3        1/1/2023       Aswan    Egypt   1007.0   7692   108.0       7.692
4        1/1/2023       Asyut    Egypt   1009.0  17731   260.0      17.731
...           ...         ...      ...      ...    ...     ...         ...
267004  3/23/2023  Sweetwater  Wyoming  56037.0  12519   139.0      12.519
267005  3/23/2023       Teton  Wyoming  56039.0  12150    16.0      12.150
267006  3/23/2023       Uinta  Wyoming  56041.0   6416    43.0       6.416
267007  3/23/2023    Washakie  Wyoming  56043.0   2700    51.0       2.700
267008  3/23/2023      Weston  Wyoming  56045.0   1906    23.0       1.906

[267009 rows x 7 columns]

In [ ]:
def classify_cases(cases):
    if cases > 50000:
        return 3
    elif cases > 20000:
        return 2
    else:
        return 1

In [ ]:
classify_cases(18961)

1

In [ ]:
df["case_severity"] = df["cases"].apply(classify_cases)

In [ ]:
df[["county", "cases", "case_severity"]].head()

county  cases  case_severity
0       Cairo  18961              1
1        Giza  67496              3
2  Alexandria   7027              1
3       Aswan   7692              1
4       Asyut  17731              1

In [ ]:
df["state_upper"] = df["state"].map(str.upper)
df[["state", "state_upper"]].head()

state state_upper
0  Egypt       EGYPT
1  Egypt       EGYPT
2  Egypt       EGYPT
3  Egypt       EGYPT
4  Egypt       EGYPT



# 🐼 **Grouping & Sorting**


# **1️⃣ Grouping in Pandas — `.groupby()`**

The **`groupby()`** method splits your data into groups based on one or more columns, then you can **aggregate**, **transform**, or **apply custom functions**.

### **Basic Syntax**

```python
df.groupby("column_name")
df.groupby(["col1", "col2"])
```

## **1.1 Group by One Column**

### Example → Total cases per state




In [ ]:
df.groupby("state")["cases"].sum()

state
Alabama           131376040
Alaska             24739745
Algeria              281319
American Samoa       672599
Arizona           195510017
                    ...    
Washington        156891026
West Virginia      52202815
Wisconsin         163131014
Wyoming            15110869
Yemen               1166192
Name: cases, Length: 78, dtype: int64




## **1.2 Group by Multiple Columns**

**total cases per county inside each state**.

In [ ]:
#pd.set_option("display.max_rows", None)
#pd.reset_option("display.max_rows")

In [ ]:
df.groupby(["state", "county"])["cases"].sum()

state    county 
Alabama  Autauga    1584709
         Baldwin    5607168
         Barbour     595568
         Bibb        646502
         Blount     1490435
                     ...   
Yemen    Marib         3566
         Saada        35041
         Sana'a       54116
         Shabwah      21981
         Taiz          4462
Name: cases, Length: 3480, dtype: int64

In [ ]:
df.groupby(["state", "county"])["cases"].mean().sort_values(ascending=False)

state       county       
California  Los Angeles      3.686937e+06
Djibouti    Dikhil           3.632440e+06
New York    New York City    3.212903e+06
Florida     Miami-Dade       1.525078e+06
Illinois    Cook             1.514331e+06
                                 ...     
Hawaii      Kalawao          1.000000e+00
Maryland    Unknown          1.000000e+00
Indiana     Unknown          1.562500e-02
Colorado    Unknown          0.000000e+00
Hawaii      Unknown          0.000000e+00
Name: cases, Length: 3480, dtype: float64


## **1.3 Using Multiple Aggregations**

Use **`.agg()`** to compute several metrics at once:


In [ ]:
df.groupby("state").agg(
    {
    "cases": ["sum", "mean", "max"],
    "deaths": ["sum", "mean"]
})

cases                             deaths             
                      sum           mean      max        sum         mean
state                                                                    
Alabama         131376040   24198.939031   239299  1733543.0   319.311660
Alaska           24739745   10908.176808   123977   115482.0    50.917989
Algeria            281319   23443.250000   121613     2874.0   239.500000
American Samoa     672599    8303.691358     8309     2754.0    34.000000
Arizona         195510017  160913.594239  1533874  2660698.0  2189.874897
...                   ...            ...      ...        ...          ...
Washington      156891026   47832.629878   553261  1273161.0   388.158841
West Virginia    52202815   11574.903548    59468   654993.0   145.231264
Wisconsin       163131014   27630.591802   343004  1330461.0   225.349085
Wyoming          15110869    8012.125663    31702   162978.0    86.414634
Yemen             1166192  106017.454545   765292    10736.0   976.000000

[78 rows x 5 columns]

## **1.4 Group & Count**

Count the number of rows per group:

In [ ]:
df.groupby("state")["county"].value_counts()

state    county 
Alabama  Autauga    81
         Baldwin    81
         Barbour    81
         Bibb       81
         Blount     81
                    ..
Yemen    Marib       1
         Saada       1
         Sana'a      1
         Shabwah     1
         Taiz        1
Name: count, Length: 3480, dtype: int64

In [ ]:
df.groupby("state")["county"].value_counts()

state    county 
Alabama  Autauga    81
         Baldwin    81
         Barbour    81
         Bibb       81
         Blount     81
                    ..
Yemen    Marib       1
         Saada       1
         Sana'a      1
         Shabwah     1
         Taiz        1
Name: count, Length: 3480, dtype: int64


## **1.5 Group & Filter**

Find states with **more than 800 total cases**:

```python
```

This returns **only rows** from California and Texas.

---

In [ ]:
df.groupby("state").filter(lambda x: x["cases"].sum() > 100000000)

date      county       state     fips   cases  deaths  \
218      1/1/2023      Placer  California   6061.0   89196   672.0   
219      1/1/2023      Plumas  California   6063.0    4240    47.0   
220      1/1/2023   Riverside  California   6065.0  759578  7863.0   
221      1/1/2023  Sacramento  California   6067.0  398826  3594.0   
222      1/1/2023  San Benito  California   6069.0   18297   110.0   
...           ...         ...         ...      ...     ...     ...   
266981  3/23/2023    Waukesha   Wisconsin  55133.0  140273  1251.0   
266982  3/23/2023     Waupaca   Wisconsin  55135.0   16431   266.0   
266983  3/23/2023    Waushara   Wisconsin  55137.0    7672    94.0   
266984  3/23/2023   Winnebago   Wisconsin  55139.0   64184   418.0   
266985  3/23/2023        Wood   Wisconsin  55141.0   27176   269.0   

        cases_in_k  case_severity state_upper  
218         89.196              3  CALIFORNIA  
219          4.240              1  CALIFORNIA  
220        759.578              3  CALIFORNIA  
221        398.826              3  CALIFORNIA  
222         18.297              1  CALIFORNIA  
...            ...            ...         ...  
266981     140.273              3   WISCONSIN  
266982      16.431              1   WISCONSIN  
266983       7.672              1   WISCONSIN  
266984      64.184              3   WISCONSIN  
266985      27.176              2   WISCONSIN  

[180182 rows x 9 columns]

## **1.6 Transforming Groups**

Sometimes, instead of collapsing groups, we want to **add new columns** based on group statistics:


In [ ]:
df["state_avg_cases"] = df.groupby("state")["cases"].transform("mean")
df

date      county    state     fips  cases  deaths  cases_in_k  \
0        1/1/2023       Cairo    Egypt   1001.0  18961   230.0      18.961   
1        1/1/2023        Giza    Egypt   1003.0  67496   719.0      67.496   
2        1/1/2023  Alexandria    Egypt   1005.0   7027   111.0       7.027   
3        1/1/2023       Aswan    Egypt   1007.0   7692   108.0       7.692   
4        1/1/2023       Asyut    Egypt   1009.0  17731   260.0      17.731   
...           ...         ...      ...      ...    ...     ...         ...   
267004  3/23/2023  Sweetwater  Wyoming  56037.0  12519   139.0      12.519   
267005  3/23/2023       Teton  Wyoming  56039.0  12150    16.0      12.150   
267006  3/23/2023       Uinta  Wyoming  56041.0   6416    43.0       6.416   
267007  3/23/2023    Washakie  Wyoming  56043.0   2700    51.0       2.700   
267008  3/23/2023      Weston  Wyoming  56045.0   1906    23.0       1.906   

        case_severity state_upper  state_avg_cases  
0                   1       EGYPT     14595.296296  
1                   3       EGYPT     14595.296296  
2                   1       EGYPT     14595.296296  
3                   1       EGYPT     14595.296296  
4                   1       EGYPT     14595.296296  
...               ...         ...              ...  
267004              1     WYOMING      8012.125663  
267005              1     WYOMING      8012.125663  
267006              1     WYOMING      8012.125663  
267007              1     WYOMING      8012.125663  
267008              1     WYOMING      8012.125663  

[267009 rows x 10 columns]

# **2️⃣ Sorting in Pandas — `.sort_values()`**

Sorting lets you organize data **by column values**.

### **Basic Syntax**

```python
df.sort_values(by="column", ascending=True/False)
```

## **2.1 Sort by One Column**

Sort states by **cases**:


In [ ]:
df.sort_values(by="cases", ascending=False)

date       county         state    fips    cases   deaths  \
263957  3/23/2023  Los Angeles    California  6037.0  3723182  35960.0   
260699  3/22/2023  Los Angeles    California  6037.0  3722448  35949.0   
257444  3/21/2023  Los Angeles    California  6037.0  3721923  35937.0   
254188  3/20/2023  Los Angeles    California  6037.0  3721276  35925.0   
247673  3/18/2023  Los Angeles    California  6037.0  3720066  35890.0   
...           ...          ...           ...     ...      ...      ...   
69029   1/22/2023      Unknown      Illinois     NaN        0   2699.0   
171367  2/22/2023      Unknown  North Dakota     NaN        0    200.0   
85173   1/27/2023      Unknown        Hawaii     NaN        0     22.0   
194173   3/1/2023      Unknown  North Dakota     NaN        0    200.0   
238251  3/15/2023      Unknown        Hawaii     NaN        0     74.0   

        cases_in_k  case_severity   state_upper  state_avg_cases  
263957    3723.182              3    CALIFORNIA    207237.904147  
260699    3722.448              3    CALIFORNIA    207237.904147  
257444    3721.923              3    CALIFORNIA    207237.904147  
254188    3721.276              3    CALIFORNIA    207237.904147  
247673    3720.066              3    CALIFORNIA    207237.904147  
...            ...            ...           ...              ...  
69029        0.000              1      ILLINOIS     39263.065712  
171367       0.000              1  NORTH DAKOTA      5265.818654  
85173        0.000              1        HAWAII     61453.686992  
194173       0.000              1  NORTH DAKOTA      5265.818654  
238251       0.000              1        HAWAII     61453.686992  

[267009 rows x 10 columns]

## **2.2 Sort by Multiple Columns**

Sort by **state**, then by **cases** within each state:

In [ ]:
df.sort_values(by=["state", "cases"], ascending=[True, False])

date     county    state    fips   cases  deaths  cases_in_k  \
260530  3/22/2023  Jefferson  Alabama  1073.0  239299  2536.0     239.299   
263788  3/23/2023  Jefferson  Alabama  1073.0  239299  2536.0     239.299   
237729  3/15/2023  Jefferson  Alabama  1073.0  239035  2535.0     239.035   
240988  3/16/2023  Jefferson  Alabama  1073.0  239035  2535.0     239.035   
244246  3/17/2023  Jefferson  Alabama  1073.0  239035  2535.0     239.035   
...           ...        ...      ...     ...     ...     ...         ...   
216      1/1/2023    Shabwah    Yemen  6057.0   21981   165.0      21.981   
210      1/1/2023   Hodeidah    Yemen  6045.0   21249   137.0      21.249   
209      1/1/2023       Taiz    Yemen  6043.0    4462    44.0       4.462   
213      1/1/2023      Marib    Yemen  6051.0    3566    17.0       3.566   
212      1/1/2023     Dhamar    Yemen  6049.0    1294    23.0       1.294   

        case_severity state_upper  state_avg_cases  
260530              3     ALABAMA     24198.939031  
263788              3     ALABAMA     24198.939031  
237729              3     ALABAMA     24198.939031  
240988              3     ALABAMA     24198.939031  
244246              3     ALABAMA     24198.939031  
...               ...         ...              ...  
216                 2       YEMEN    106017.454545  
210                 2       YEMEN    106017.454545  
209                 1       YEMEN    106017.454545  
213                 1       YEMEN    106017.454545  
212                 1       YEMEN    106017.454545  

[267009 rows x 10 columns]

In [ ]:
df.set_index("date")

county    state     fips  cases  deaths  cases_in_k  \
date                                                                 
1/1/2023        Cairo    Egypt   1001.0  18961   230.0      18.961   
1/1/2023         Giza    Egypt   1003.0  67496   719.0      67.496   
1/1/2023   Alexandria    Egypt   1005.0   7027   111.0       7.027   
1/1/2023        Aswan    Egypt   1007.0   7692   108.0       7.692   
1/1/2023        Asyut    Egypt   1009.0  17731   260.0      17.731   
...               ...      ...      ...    ...     ...         ...   
3/23/2023  Sweetwater  Wyoming  56037.0  12519   139.0      12.519   
3/23/2023       Teton  Wyoming  56039.0  12150    16.0      12.150   
3/23/2023       Uinta  Wyoming  56041.0   6416    43.0       6.416   
3/23/2023    Washakie  Wyoming  56043.0   2700    51.0       2.700   
3/23/2023      Weston  Wyoming  56045.0   1906    23.0       1.906   

           case_severity state_upper  state_avg_cases  
date                                                   
1/1/2023               1       EGYPT     14595.296296  
1/1/2023               3       EGYPT     14595.296296  
1/1/2023               1       EGYPT     14595.296296  
1/1/2023               1       EGYPT     14595.296296  
1/1/2023               1       EGYPT     14595.296296  
...                  ...         ...              ...  
3/23/2023              1     WYOMING      8012.125663  
3/23/2023              1     WYOMING      8012.125663  
3/23/2023              1     WYOMING      8012.125663  
3/23/2023              1     WYOMING      8012.125663  
3/23/2023              1     WYOMING      8012.125663  

[267009 rows x 9 columns]

# **3️⃣ Combining Grouping + Sorting**

### Example → Top 2 counties per state based on cases:


In [ ]:
df.groupby("state").apply(lambda x: x.sort_values("cases", ascending=False).head(2))

C:\Users\Nima\AppData\Local\Temp\ipykernel_45696\3055627841.py:1: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  df.groupby("state").apply(lambda x: x.sort_values("cases", ascending=False).head(2))


date     county      state     fips   cases  deaths  \
state                                                                        
Alabama   263788  3/23/2023  Jefferson    Alabama   1073.0  239299  2536.0   
          260530  3/22/2023  Jefferson    Alabama   1073.0  239299  2536.0   
Alaska    257308  3/21/2023  Anchorage     Alaska   2020.0  123977   593.0   
          263821  3/23/2023  Anchorage     Alaska   2020.0  123977   593.0   
Algeria   70       1/1/2023     Skikda    Algeria   2020.0  121613   579.0   
...                     ...        ...        ...      ...     ...     ...   
Wisconsin 263697  3/22/2023  Milwaukee  Wisconsin  55079.0  342846  2502.0   
Wyoming   263739  3/22/2023    Laramie    Wyoming  56021.0   31702   322.0   
          266996  3/23/2023    Laramie    Wyoming  56021.0   31702   322.0   
Yemen     217      1/1/2023   Al Bayda      Yemen   6059.0  765292  7674.0   
          214      1/1/2023  Hadramaut      Yemen   6053.0  118770   803.0   

                  cases_in_k  case_severity state_upper  state_avg_cases  
state                                                                     
Alabama   263788     239.299              3     ALABAMA     24198.939031  
          260530     239.299              3     ALABAMA     24198.939031  
Alaska    257308     123.977              3      ALASKA     10908.176808  
          263821     123.977              3      ALASKA     10908.176808  
Algeria   70         121.613              3     ALGERIA     23443.250000  
...                      ...            ...         ...              ...  
Wisconsin 263697     342.846              3   WISCONSIN     27630.591802  
Wyoming   263739      31.702              2     WYOMING      8012.125663  
          266996      31.702              2     WYOMING      8012.125663  
Yemen     217        765.292              3       YEMEN    106017.454545  
          214        118.770              3       YEMEN    106017.454545  

[156 rows x 10 columns]

In [ ]:
# Total deaths per state sorted descending
df.groupby("state")["deaths"].sum().sort_values(ascending=False)

state
California      8366946.0
Texas           7679552.0
Florida         7020942.0
New York        6483341.0
Pennsylvania    4093547.0
                  ...    
Lebanon            1067.0
Palestine          1007.0
Tunisia             698.0
Qatar               457.0
Libya               106.0
Name: deaths, Length: 78, dtype: float64

In [ ]:
# Top 5 states with highest avg cases
df.groupby("state")["deaths"].mean().sort_values(ascending=False)

state
Djibouti                    7596.200000
Puerto Rico                 5721.378049
Arizona                     2189.874897
California                  1770.407533
Iraq                        1756.888889
                               ...     
North Dakota                  46.280939
Virgin Islands                36.136519
American Samoa                34.000000
Northern Mariana Islands      13.666667
Libya                         11.777778
Name: deaths, Length: 78, dtype: float64

# **Summary Table**

| **Task**              | **Function**     | **Example**                                                  |
| --------------------- | ---------------- | ------------------------------------------------------------ |
| Group data            | `.groupby()`     | `df.groupby("state")["cases"].sum()`                         |
| Multiple aggregations | `.agg()`         | `df.groupby("state").agg({"cases":["sum","mean"]})`          |
| Count groups          | `.count()`       | `df.groupby("state")["county"].count()`                      |
| Filter groups         | `.filter()`      | `df.groupby("state").filter(lambda x: x["cases"].sum()>800)` |
| Transform values      | `.transform()`   | `df["avg"]=df.groupby("state")["cases"].transform("mean")`   |
| Sort data             | `.sort_values()` | `df.sort_values("cases",ascending=False)`                    |

---

# **Next Steps**

Now you know:

* ✅ Grouping (`groupby`, `agg`, `transform`)
* ✅ Sorting (`sort_values`, `sort_index`)
* ✅ Combining grouping + sorting
* ✅ Advanced filtering per group



### **Checking for Missing Data**


In [ ]:
pd.isnull(df["cases"])

0         False
1         False
2         False
3         False
4         False
          ...  
267004    False
267005    False
267006    False
267007    False
267008    False
Name: cases, Length: 267009, dtype: bool

## **2️⃣ Counting Missing Values**

To **count** missing values per column:



In [ ]:
df.isnull().sum()

date                  0
county                0
state                 0
fips               2986
cases                 0
deaths             6396
cases_in_k            0
case_severity         0
state_upper           0
state_avg_cases       0
dtype: int64

### **Find all rows with NaN in `cases`:**

```python


In [ ]:
df[df["deaths"].isnull()]

date        county        state     fips  cases  deaths  \
2335     1/1/2023      Adjuntas  Puerto Rico  72001.0   6551     NaN   
2336     1/1/2023        Aguada  Puerto Rico  72003.0  12524     NaN   
2337     1/1/2023     Aguadilla  Puerto Rico  72005.0  16023     NaN   
2338     1/1/2023  Aguas Buenas  Puerto Rico  72007.0   8229     NaN   
2339     1/1/2023      Aibonito  Puerto Rico  72009.0   8191     NaN   
...           ...           ...          ...      ...    ...     ...   
266162  3/23/2023     Vega Baja  Puerto Rico  72145.0  20961     NaN   
266163  3/23/2023       Vieques  Puerto Rico  72147.0   1891     NaN   
266164  3/23/2023      Villalba  Puerto Rico  72149.0   7804     NaN   
266165  3/23/2023       Yabucoa  Puerto Rico  72151.0  10631     NaN   
266166  3/23/2023         Yauco  Puerto Rico  72153.0  11826     NaN   

        cases_in_k  case_severity  state_upper  state_avg_cases  
2335         6.551              1  PUERTO RICO     14174.394721  
2336        12.524              1  PUERTO RICO     14174.394721  
2337        16.023              1  PUERTO RICO     14174.394721  
2338         8.229              1  PUERTO RICO     14174.394721  
2339         8.191              1  PUERTO RICO     14174.394721  
...            ...            ...          ...              ...  
266162      20.961              2  PUERTO RICO     14174.394721  
266163       1.891              1  PUERTO RICO     14174.394721  
266164       7.804              1  PUERTO RICO     14174.394721  
266165      10.631              1  PUERTO RICO     14174.394721  
266166      11.826              1  PUERTO RICO     14174.394721  

[6396 rows x 10 columns]

In [ ]:
df[df["deaths"].notnull()]

date      county    state     fips  cases  deaths  cases_in_k  \
0        1/1/2023       Cairo    Egypt   1001.0  18961   230.0      18.961   
1        1/1/2023        Giza    Egypt   1003.0  67496   719.0      67.496   
2        1/1/2023  Alexandria    Egypt   1005.0   7027   111.0       7.027   
3        1/1/2023       Aswan    Egypt   1007.0   7692   108.0       7.692   
4        1/1/2023       Asyut    Egypt   1009.0  17731   260.0      17.731   
...           ...         ...      ...      ...    ...     ...         ...   
267004  3/23/2023  Sweetwater  Wyoming  56037.0  12519   139.0      12.519   
267005  3/23/2023       Teton  Wyoming  56039.0  12150    16.0      12.150   
267006  3/23/2023       Uinta  Wyoming  56041.0   6416    43.0       6.416   
267007  3/23/2023    Washakie  Wyoming  56043.0   2700    51.0       2.700   
267008  3/23/2023      Weston  Wyoming  56045.0   1906    23.0       1.906   

        case_severity state_upper  state_avg_cases  
0                   1       EGYPT     14595.296296  
1                   3       EGYPT     14595.296296  
2                   1       EGYPT     14595.296296  
3                   1       EGYPT     14595.296296  
4                   1       EGYPT     14595.296296  
...               ...         ...              ...  
267004              1     WYOMING      8012.125663  
267005              1     WYOMING      8012.125663  
267006              1     WYOMING      8012.125663  
267007              1     WYOMING      8012.125663  
267008              1     WYOMING      8012.125663  

[260613 rows x 10 columns]

### **Fill with the Column Mean**


In [ ]:
df["deaths"].mean()

np.float64(352.2099242938764)

In [ ]:
df["deaths"].fillna(df["deaths"].mean(), inplace=True)

C:\Users\Nima\AppData\Local\Temp\ipykernel_45696\3618912776.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df["deaths"].fillna(df["deaths"].mean(), inplace=True)


In [ ]:
df

date      county    state     fips  cases  deaths  cases_in_k  \
0        1/1/2023       Cairo    Egypt   1001.0  18961   230.0      18.961   
1        1/1/2023        Giza    Egypt   1003.0  67496   719.0      67.496   
2        1/1/2023  Alexandria    Egypt   1005.0   7027   111.0       7.027   
3        1/1/2023       Aswan    Egypt   1007.0   7692   108.0       7.692   
4        1/1/2023       Asyut    Egypt   1009.0  17731   260.0      17.731   
...           ...         ...      ...      ...    ...     ...         ...   
267004  3/23/2023  Sweetwater  Wyoming  56037.0  12519   139.0      12.519   
267005  3/23/2023       Teton  Wyoming  56039.0  12150    16.0      12.150   
267006  3/23/2023       Uinta  Wyoming  56041.0   6416    43.0       6.416   
267007  3/23/2023    Washakie  Wyoming  56043.0   2700    51.0       2.700   
267008  3/23/2023      Weston  Wyoming  56045.0   1906    23.0       1.906   

        case_severity state_upper  state_avg_cases  
0                   1       EGYPT     14595.296296  
1                   3       EGYPT     14595.296296  
2                   1       EGYPT     14595.296296  
3                   1       EGYPT     14595.296296  
4                   1       EGYPT     14595.296296  
...               ...         ...              ...  
267004              1     WYOMING      8012.125663  
267005              1     WYOMING      8012.125663  
267006              1     WYOMING      8012.125663  
267007              1     WYOMING      8012.125663  
267008              1     WYOMING      8012.125663  

[267009 rows x 10 columns]

In [ ]:
pd.isnull(df["deaths"])

0         False
1         False
2         False
3         False
4         False
          ...  
267004    False
267005    False
267006    False
267007    False
267008    False
Name: deaths, Length: 267009, dtype: bool

In [ ]:
df["cases"].median()

7955.0

In [ ]:
df["cases"].fillna(df["cases"].median(), inplace=True)

C:\Users\Nima\AppData\Local\Temp\ipykernel_45696\3926648164.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df["cases"].fillna(df["cases"].median(), inplace=True)


### **Fill Missing Values by Group (per state)**

For example, fill missing `fips` **within each state** using the **state-wise mean**:


This ensures that missing values in each **state** column are filled using **only that state's average**.


In [ ]:
df["deaths"] = df.groupby("state")["deaths"].transform(lambda x: x.fillna(x.mean()))

## **5️⃣ Dropping Missing Values**

If you want to **remove rows** that contain NaN:

```python
df.dropna(inplace=True)
```

Or remove rows **only if all columns are NaN**:

```python
df.dropna(how="all", inplace=True)
```

In [ ]:
df.dropna(inplace=True)

In [ ]:
df

date      county    state     fips  cases  deaths  cases_in_k  \
0        1/1/2023       Cairo    Egypt   1001.0  18961   230.0      18.961   
1        1/1/2023        Giza    Egypt   1003.0  67496   719.0      67.496   
2        1/1/2023  Alexandria    Egypt   1005.0   7027   111.0       7.027   
3        1/1/2023       Aswan    Egypt   1007.0   7692   108.0       7.692   
4        1/1/2023       Asyut    Egypt   1009.0  17731   260.0      17.731   
...           ...         ...      ...      ...    ...     ...         ...   
267004  3/23/2023  Sweetwater  Wyoming  56037.0  12519   139.0      12.519   
267005  3/23/2023       Teton  Wyoming  56039.0  12150    16.0      12.150   
267006  3/23/2023       Uinta  Wyoming  56041.0   6416    43.0       6.416   
267007  3/23/2023    Washakie  Wyoming  56043.0   2700    51.0       2.700   
267008  3/23/2023      Weston  Wyoming  56045.0   1906    23.0       1.906   

        case_severity state_upper  state_avg_cases  
0                   1       EGYPT     14595.296296  
1                   3       EGYPT     14595.296296  
2                   1       EGYPT     14595.296296  
3                   1       EGYPT     14595.296296  
4                   1       EGYPT     14595.296296  
...               ...         ...              ...  
267004              1     WYOMING      8012.125663  
267005              1     WYOMING      8012.125663  
267006              1     WYOMING      8012.125663  
267007              1     WYOMING      8012.125663  
267008              1     WYOMING      8012.125663  

[264023 rows x 10 columns]

In [ ]:
df.dropna(how="all", inplace=True)


## **Summary Table**

| **Method**                 | **Use Case**                       |
| -------------------------- | ---------------------------------- |
| `pd.isnull()`              | Check for NaN values               |
| `pd.notnull()`             | Check for non-NaN values           |
| `df.isnull().sum()`        | Count missing values per column    |
| `df.fillna(x)`             | Replace NaN with a fixed value `x` |
| `df.fillna(df.mean())`     | Replace NaN with column mean       |
| `df.groupby().transform()` | Fill NaN within each group         |
| `df.dropna()`              | Remove rows containing NaN         |

---

## **📝 Exercises**
1. Find the **average number of deaths** per state.
2. List all counties in **California**.
3. Find the **top 5 states** with the highest number of cases.
4. Create a new column showing `death_rate = deaths / cases`.
---



# 🐼Renaming Columns & Combining DataFrames**

---

## **📌 1. Renaming Columns**

In Pandas, we use **`.rename()`** to **change column names** or **index names**.
This is helpful when your dataset has unclear, inconsistent, or unwanted names.

---
### **🔹 Rename One Column**

In [ ]:
df.rename(columns={"cases": "total_cases"}, inplace=True)
df.head()

date      county  state    fips  total_cases  deaths  cases_in_k  \
0  1/1/2023       Cairo  Egypt  1001.0        18961   230.0      18.961   
1  1/1/2023        Giza  Egypt  1003.0        67496   719.0      67.496   
2  1/1/2023  Alexandria  Egypt  1005.0         7027   111.0       7.027   
3  1/1/2023       Aswan  Egypt  1007.0         7692   108.0       7.692   
4  1/1/2023       Asyut  Egypt  1009.0        17731   260.0      17.731   

   case_severity state_upper  state_avg_cases  
0              1       EGYPT     14595.296296  
1              3       EGYPT     14595.296296  
2              1       EGYPT     14595.296296  
3              1       EGYPT     14595.296296  
4              1       EGYPT     14595.296296

### **🔹 Rename Multiple Columns**

In [ ]:
df.rename(columns={
    "date": "reported_date",
    "deaths": "total_deaths"
}, inplace=True)

In [ ]:
df.head()

reported_date      county  state    fips  total_cases  total_deaths  \
0      1/1/2023       Cairo  Egypt  1001.0        18961         230.0   
1      1/1/2023        Giza  Egypt  1003.0        67496         719.0   
2      1/1/2023  Alexandria  Egypt  1005.0         7027         111.0   
3      1/1/2023       Aswan  Egypt  1007.0         7692         108.0   
4      1/1/2023       Asyut  Egypt  1009.0        17731         260.0   

   cases_in_k  case_severity state_upper  state_avg_cases  
0      18.961              1       EGYPT     14595.296296  
1      67.496              3       EGYPT     14595.296296  
2       7.027              1       EGYPT     14595.296296  
3       7.692              1       EGYPT     14595.296296  
4      17.731              1       EGYPT     14595.296296


### **🔹 Change All Column Names at Once**

In [ ]:
df.columns = ["Date", "County", "State", "FIPS", "Cases", "Deaths"]
df.head()

Date      County  State    FIPS  Cases  Deaths
0  1/1/2023       Cairo  Egypt  1001.0  18961   230.0
1  1/1/2023        Giza  Egypt  1003.0  67496   719.0
2  1/1/2023  Alexandria  Egypt  1005.0   7027   111.0
3  1/1/2023       Aswan  Egypt  1007.0   7692   108.0
4  1/1/2023       Asyut  Egypt  1009.0  17731   260.0

### **📝 Quick Exercise 1**

🔹 Rename the following columns:

* `cases` → `COVID_Cases`
* `deaths` → `COVID_Deaths`

<details>
<summary>💡 Hint</summary>

```python
df.rename(columns={"cases": "COVID_Cases", "deaths": "COVID_Deaths"}, inplace=True)
```

</details>

## **📌 2. Combining DataFrames**

Sometimes, you have data in **multiple DataFrames** and want to **merge** or **concatenate** them.

### **🔹 Example: Combine Daily Data**

Let's create **two DataFrames** for **January 1st** and **January 2nd**:

In [ ]:
df_day1 = df[df["Date"] == "1/1/2023"]
df_day2 = df[df["Date"] == "1/2/2023"]

print(df_day1)
print(df_day2)

          Date      County    State     FIPS  Cases  Deaths
0     1/1/2023       Cairo    Egypt   1001.0  18961   230.0
1     1/1/2023        Giza    Egypt   1003.0  67496   719.0
2     1/1/2023  Alexandria    Egypt   1005.0   7027   111.0
3     1/1/2023       Aswan    Egypt   1007.0   7692   108.0
4     1/1/2023       Asyut    Egypt   1009.0  17731   260.0
...        ...         ...      ...      ...    ...     ...
3250  1/1/2023  Sweetwater  Wyoming  56037.0  12410   136.0
3251  1/1/2023       Teton  Wyoming  56039.0  12010    16.0
3252  1/1/2023       Uinta  Wyoming  56041.0   6305    43.0
3253  1/1/2023    Washakie  Wyoming  56043.0   2721    47.0
3254  1/1/2023      Weston  Wyoming  56045.0   1880    22.0

[3255 rows x 6 columns]
          Date      County    State     FIPS  Cases  Deaths
3255  1/2/2023     Autauga  Alabama   1001.0  18961   230.0
3256  1/2/2023     Baldwin  Alabama   1003.0  67496   719.0
3257  1/2/2023     Barbour  Alabama   1005.0   7027   111.0
3258  1/2/2023 

### **🔹 Concatenate Data (Stack Rows)**

In [ ]:
df_combined = pd.concat([df_day1, df_day2])
df_combined

Date      County    State     FIPS  Cases  Deaths
0     1/1/2023       Cairo    Egypt   1001.0  18961   230.0
1     1/1/2023        Giza    Egypt   1003.0  67496   719.0
2     1/1/2023  Alexandria    Egypt   1005.0   7027   111.0
3     1/1/2023       Aswan    Egypt   1007.0   7692   108.0
4     1/1/2023       Asyut    Egypt   1009.0  17731   260.0
...        ...         ...      ...      ...    ...     ...
6505  1/2/2023  Sweetwater  Wyoming  56037.0  12410   136.0
6506  1/2/2023       Teton  Wyoming  56039.0  12010    16.0
6507  1/2/2023       Uinta  Wyoming  56041.0   6305    43.0
6508  1/2/2023    Washakie  Wyoming  56043.0   2721    47.0
6509  1/2/2023      Weston  Wyoming  56045.0   1880    22.0

[6510 rows x 6 columns]

### **🔹 Merge DataFrames (Join on Column)**

If you have two DataFrames with **shared columns**, you can use **`.merge()`**:

In [ ]:
df_population = pd.DataFrame({
    "County": ["Autauga", "Baldwin", "Barbour"],
    "population": [55869, 227131, 24686]
})
df_population

County  population
0  Autauga       55869
1  Baldwin      227131
2  Barbour       24686

In [ ]:
df_merged = pd.merge(df, df_population, on="County", how="left")
df_merged

Date      County    State     FIPS  Cases  Deaths  population
0        1/1/2023       Cairo    Egypt   1001.0  18961   230.0         NaN
1        1/1/2023        Giza    Egypt   1003.0  67496   719.0         NaN
2        1/1/2023  Alexandria    Egypt   1005.0   7027   111.0         NaN
3        1/1/2023       Aswan    Egypt   1007.0   7692   108.0         NaN
4        1/1/2023       Asyut    Egypt   1009.0  17731   260.0         NaN
...           ...         ...      ...      ...    ...     ...         ...
267004  3/23/2023  Sweetwater  Wyoming  56037.0  12519   139.0         NaN
267005  3/23/2023       Teton  Wyoming  56039.0  12150    16.0         NaN
267006  3/23/2023       Uinta  Wyoming  56041.0   6416    43.0         NaN
267007  3/23/2023    Washakie  Wyoming  56043.0   2700    51.0         NaN
267008  3/23/2023      Weston  Wyoming  56045.0   1906    23.0         NaN

[267009 rows x 7 columns]


# 🔹 **Understanding `how` in `pd.merge()`**

When we merge two **DataFrames** in Pandas using `pd.merge()`, the **`how`** parameter **controls the type of join** — just like **SQL joins**.

---

## **📌 Syntax**

```python
pd.merge(left_df, right_df, on="common_column", how="type_of_join")
```

* **`left_df`** → First DataFrame
* **`right_df`** → Second DataFrame
* **`on`** → Column name to merge on
* **`how`** → Type of join (controls which rows are included)

---

## **🔹 Types of Joins (`how`)**

| **Type**              | **Keeps Rows From**         | **Keeps Unmatched from Left** | **Keeps Unmatched from Right** | **Result Size**           |
| --------------------- | --------------------------- | ----------------------------- | ------------------------------ | ------------------------- |
| **inner** *(default)* | Only matching rows          | ❌ No                          | ❌ No                           | Smallest                  |
| **left**              | All rows from **left\_df**  | ✅ Yes                         | ❌ No                           | At least as many as left  |
| **right**             | All rows from **right\_df** | ❌ No                          | ✅ Yes                          | At least as many as right |
| **outer**             | All rows from **both**      | ✅ Yes                         | ✅ Yes                          | Largest                   |

---

## **🔹 Example DataFrames**

### **Main Data (df)**

| county  | cases |
| ------- | ----- |
| Autauga | 18961 |
| Baldwin | 67496 |
| Barbour | 7027  |
| Bibb    | 7692  |

### **Population Data (df\_population)**

| county  | population |
| ------- | ---------- |
| Autauga | 55869      |
| Baldwin | 227131     |
| Blount  | 59852      |

---

## **1️⃣ Inner Join (Default)**

```python
df_merged = pd.merge(df, df_population, on="county", how="inner")
print(df_merged)
```

**Result:**

| county  | cases | population |
| ------- | ----- | ---------- |
| Autauga | 18961 | 55869      |
| Baldwin | 67496 | 227131     |

✅ Keeps **only counties present in both DataFrames**.

---

## **2️⃣ Left Join**

```python
df_merged = pd.merge(df, df_population, on="county", how="left")
print(df_merged)
```

**Result:**

| county  | cases | population |
| ------- | ----- | ---------- |
| Autauga | 18961 | 55869      |
| Baldwin | 67496 | 227131     |
| Barbour | 7027  | NaN        |
| Bibb    | 7692  | NaN        |

✅ Keeps **all rows from `df` (left)**.
❌ If a county is **missing in `df_population`**, we get **NaN**.

---

## **3️⃣ Right Join**

```python
df_merged = pd.merge(df, df_population, on="county", how="right")
print(df_merged)
```

**Result:**

| county  | cases | population |
| ------- | ----- | ---------- |
| Autauga | 18961 | 55869      |
| Baldwin | 67496 | 227131     |
| Blount  | NaN   | 59852      |

✅ Keeps **all rows from `df_population` (right)**.
❌ If a county is **missing in `df`**, `cases` will be **NaN**.

---

## **4️⃣ Outer Join**

```python
df_merged = pd.merge(df, df_population, on="county", how="outer")
print(df_merged)
```

**Result:**

| county  | cases | population |
| ------- | ----- | ---------- |
| Autauga | 18961 | 55869      |
| Baldwin | 67496 | 227131     |
| Barbour | 7027  | NaN        |
| Bibb    | 7692  | NaN        |
| Blount  | NaN   | 59852      |

✅ Keeps **all rows from both DataFrames**.
❌ Missing values become **NaN**.

---

## **📌 Summary of `how`**

| **how** | **Keeps all from Left** | **Keeps all from Right** | **Keeps only Matches** |
| ------- | ----------------------- | ------------------------ | ---------------------- |
| inner   | ❌                       | ❌                        | ✅                      |
| left    | ✅                       | ❌                        | ✅                      |
| right   | ❌                       | ✅                        | ✅                      |
| outer   | ✅                       | ✅                        | ✅                      |

---



## ** 1. Overview**

| **Feature**          | **`pd.concat()`** 🟢                                                                   | **`pd.merge()`** 🔵                                                          |
| -------------------- | -------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------- |
| **Purpose**          | Combines **DataFrames** **on top of each other** (rows) or **side by side** (columns). | Combines DataFrames **based on common columns or indexes** (like SQL joins). |
| **Join Type**        | Simple stacking, **no matching needed** unless specified.                              | Smart combining, **matches rows based on keys**.                             |
| **Default Behavior** | Vertical stacking (adds rows).                                                         | Inner join (keeps matching rows).                                            |
| **Axis**             | You can specify `axis=0` (rows, default) or `axis=1` (columns).                        | Always merges **based on a key**, not axes.                                  |
| **Use Case**         | Adding **new rows** or **new columns** when structure is similar.                      | Combining **related datasets** using a common key.                           |

---

## ** 2. Using `pd.concat()`** → **Stacking Data**


### **🔹 Example 1 — Vertical Concatenation (Add Rows)**

```python
df_concat = pd.concat([df1, df2], axis=0)
print(df_concat)
```


### **🔹 Example 2 — Horizontal Concatenation (Add Columns)**
```python
df_concat_cols = pd.concat([df1, df3], axis=1)
print(df_concat_cols)
```

## **🔹  Using `pd.merge()` ** → **Smart Combining Based on Keys**


### **🔹 Example 3 — Merge on a Common Column**

```python
df_merged = pd.merge(df1, df_population, on="county", how="left")
print(df_merged)
```


## **🔹  When to Use What**

* ✅ **Use `concat()`** when you just want to **append rows** or **stack columns**.
* ✅ **Use `merge()`** when you want to **combine datasets intelligently** based on **common columns** (like SQL joins).

---





## **📌 4. Exercises for You**

### **Exercise 1** 📝

Rename:

* `fips` → `FIPS_Code`
* `county` → `County_Name`

---

### **Exercise 2** 📝

Create two DataFrames:

* One for **1/1/2023**
* One for **1/2/2023**
  Then combine them using **`pd.concat()`**.

---

### **Exercise 3** 📝

Merge the dataset with a new DataFrame that contains:

* `county`
* `vaccinated_people`

Use `pd.merge()` to combine them.




<div style="background-color:#28a745; padding:15px; border-radius:10px; text-align:center; display:flex; justify-content:center; align-items:center; gap:12px;">
    <img src="attachment:31a19788-73b0-412f-a280-6903c16f228c.jpg" alt="Evergreen Logo" style="width:70px; height:70px; border-radius:50%;">
    <h1 style="color:white; font-size:40px; font-weight:bold; margin:0;">
        Evergreen.ai
    </h1>

<div style="margin-left:auto; padding-right:10px;">
        <h2 style="color:white; font-size:30px; font-weight:600; margin:0;">
            Eng / Mahmoud Talaat ->
            Thanks
        </h2>
        <h2 style="color:white; font-size:30px; font-weight:600; margin:0;">
            01146544662
        </h2>
</div>
</div>
